### PR TITLE
Restore -static-executable on Linux.

### PR DESCRIPTION
**Pull request by spevans**
- Build libswiftImageInspectionStatic.a from
  ImageInspectionELF.cpp and StaticBinaryELF.cpp

- Update static-executable-args.lnk and replace
  libswiftImageInspectionShared with libswiftImageInspectionStatic.

This fixes `-static-executable` on Linux to build a static binary (when linked with appropiate libicu .a libs`

### CODE REVIEW

**Comment by compnerd** [comment]
Hmm, why not leave this as is, and just modify the swiftrt.o to have a static version which doesn't invoke the register method?  That way, this can fully be reincorporated into the runtime, and the driver can make the appropriate choice.

**Review by compnerd — DISMISSED** [review]
See previous comment, I suppose that the notification might work better this way.

**Comment by spevans** [comment]
@compnerd Sorry for the delay in replying, are you suggesting something like create a `swiftrt-static.o` which doesn't have the call to `swift_addNewDSOImage` and then modify `lib/Driver/ToolChains.cpp` to link that version instead, when a static executable is built?

**Comment by compnerd** [comment]
@spevans, yes, effectively that.  However, thinking more about this, we can be even more clever about this.  On the static builds, you still need to link to the (static) runtime.  The static runtime can include the object file, and we can create a reference to it, causing the linker to preserve the object from the runtime itself and avoid having to have the driver push the file into the actual linker invocation.

**Comment by spevans** [comment]
@compnerd In the most recent version I created `ImageInspectionStatic-ELF.cpp` which is effectively `SwiftRT-ELF.cpp` with the constructor and `swift_addNewDSOImage` removed. I also removed linking `SwiftRT.o` from `lib/Driver/Toolchains.cpp` for static executables. I think this matches what you are suggesting.

**Comment by compnerd** [comment]
While functionally equivalent, this introduces yet another copy of the table of sections.  It is another place where we need to update that list.  If we compile the single file with a macro we can get away with a single place for all of ELF, which I think is better.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci please test

File: stdlib/public/runtime/SwiftRT-ELF.cpp [hunk 1]
```diff
@@ -63,8 +66,50 @@ static void swift_image_constructor() {
...
+  void initializeTypeFieldLookup() {
+    const MetadataSections::Range fields_range =
+      SWIFT_SECTION_RANGE(swift4_fieldmd);
+    if (fields_range.length) {
+      const void *start = reinterpret_cast<void *>(fields_range.start);
+      addImageTypeFieldDescriptorBlockCallback(start, fields_range.length);
+    }
+  }
+}
+
```
Comment (compnerd) [inline stdlib/public/runtime/SwiftRT-ELF.cpp]: Isn't this just a duplication of the `swift_addNewDSOImage`?  Why not just use that?

**Review by compnerd — COMMENTED** [review]
I think that it might be possible to reduce some duplication of the code which would be better.

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci test linux

**Comment by spevans** [comment]
@swift-ci Please smoke test OS X platform

**Comment by spevans** [comment]
@compnerd I know this PR is quite old but I have updated it including a test and removed the duplicate sections I added in `SwfitRT-ELF.cpp`. Could you take another look at it please?

**Comment by frol** [comment]
I have bumped into this issue and filed a bug: https://bugs.swift.org/browse/SR-10328.

**Comment by spevans** [comment]
@frol Unfortunately this PR still has some issues on Ubuntu18.04, relating to glibc and pthreads so it still needs some more work

**Comment by jckarter** [comment]
> While functionally equivalent, this introduces yet another copy of the table of sections. It is another place where we need to update that list. If we compile the single file with a macro we can get away with a single place for all of ELF, which I think is better.

As I think you were alluding to, in a fully-static binary there really isn't any need for a table of sections at all, since we ought to be able to link the runtime directly against the beginnings and ends of the sections from the main program.

**Comment by miguelangel-dev** [comment]
Just a quick question, I was using `-static-executable` in macox, with sdk `iphonesimulator` and target `x86_64-apple-ios12.2-simulator`- it stops to work (I think from Swift 5) and it throws the next error:
```
llvm error: -static-executable is not supported on darwin
```

This PR will solve this error?

**Comment by spevans** [comment]
Apple platforms do not support statically linked executables, see https://developer.apple.com/library/content/qa/qa1118/_index.html

**Comment by spevans** [comment]
Please test with following PR:

apple/swift-integration-tests#59
@swift-ci test

**Comment by spevans** [comment]
Please test with following PR:

apple/swift-integration-tests#59
@swift-ci test

**Comment by spevans** [comment]
Please test with following PR:

apple/swift-integration-tests#59
@swift-ci test linux

**Comment by CodaFi** [comment]
@spevans It's been a while and this has gathered quite a few conflicts. Is this still something you're interested in pursuing?

**Comment by aemino** [comment]
Would restoring the static executable flag for Linux allow statically-linked toolchains on Linux? Because if so, I think that would be incredibly helpful for producing cross-distro compatible Linux toolchains.

**Comment by spevans** [comment]
@CodaFi I'll close this off for now, I need to get `--static-stdlib` working with Foundation first anyway and this PR had issues on Ubutnu18.04 if I recall (it worked ok on 16.04)

@aemino This PR was for producing statically linked binaries as the output of `swiftc` not actually building statically linked versions of the compiler itself.

**Pull request closed without merging**